### PR TITLE
Whammy Pitch Modulation

### PR DESCRIPTION
**Pull request by jbienz**
Added pitch bending whammy support back in.

KFfitLK6Ac

I have updated this PR to move the Whammy settings under Experimental Features. This means that the following PR would need to be merged before this one can be merged:

https://github.com/YARC-Official/YARG/pull/815

I submitted another PR for YARG.Core to support these changes and I will open a discussion in this feature request:

https://github.com/YARC-Official/YARG/issues/786

Here are the changes I implemented to restore this feature:

1. Added WhammyPitch (float) to StemSettings

2. Added OnWhammyPitchChange to StemSettings

3. In StemChannel subscribed to OnWhammyPitchChange at the same location volume change is called

4. In GlobalAudioHandler added Get/SetWhammyPitchSetting 

5. In StemState added SetWhammyPitch.

6. In SettingsManager.Settings uncommented UseWhammyFx setting and UseWhammyFxChange.

7. In SettingsManager.Settings uncommented WhammyPitchShiftAmount setting and WhammyPitchShiftAmountChange.

8. In GameManager.Audio added ChangeStemWhammyPitch

9. In FiveFretPlayer added SetStemWhammyPitch

10. In FiveFretPlayer called SetStemWhammyPitch in override for OnInputQueued when WhammyFactor is updated.

11. In GlobalAudioHandler changed WHAMMY_FFT_DEFAULT to 512 to try and reduce latency between guitar track and other tracks when pitch change is enabled.

### CODE REVIEW

**Comment by jbienz** [comment]
I have updated this PR to move the Whammy settings under Experimental Features. This means that the following PR would need to be merged before this one can be merged:

https://github.com/YARC-Official/YARG/pull/815

**Comment by jbienz** [comment]
Updated to incorporate changes made in https://github.com/YARC-Official/YARG/pull/815.

File: Assets/Script/Gameplay/GameManager.Audio.cs [hunk 1]
```diff
@@ -49,6 +50,14 @@ public bool SetReverb(bool reverb)
...
             }
 
+            public float SetWhammyPitch(float percent)
+            {
+                // TODO: Would be nice to handle multiple inputs
+                // but for now last one wins
+                WhammyPitch = (percent < 0 || percent > 1 ? 0 : percent);
+                return WhammyPitch;
+            }
+
```
Comment (sonicfind) [inline Assets/Script/Gameplay/GameManager.Audio.cs:60]: With my recommendation of disallowing multiple players to share a stem for whammying purposes, this function and the variable wouldn't need to be added.

**Review by sonicfind — CHANGES_REQUESTED** [review]

**Comment by sonicfind** [comment]
Heads up, I've mentioned this to a few others, but don't use merge when you're handling a PR. Always **rebase**. If you merge, it becomes hard to follow some of the flow of changes made through each commit as it fills the branch with nondescript updates. Yours isn't as bad as others, but you do still have five merge commits in this PR.

File: Assets/Script/Gameplay/GameManager.Audio.cs [hunk 2]
```diff
@@ -156,5 +165,33 @@ public void ChangeStemReverbState(SongStem stem, bool reverb)
...
+            // Whammy is not supported on single track
+            if (stem == _backgroundStem)
+            {
+                return;
+            }
+
+            float percentActive = state.SetWhammyPitch(percent);
+            GlobalAudioHandler.SetWhammyPitchSetting(stem, percentActive);
+        }
```
[the remark this answers is not in the document]
Comment (jbienz) [reply Assets/Script/Gameplay/GameManager.Audio.cs:195]: Yes, it's certainly possible for more than one player to be Guitar or Bass. I would still like to allow it, even if there are multiple guitarists. This happens often at our parties. Perhaps that's another option?

- Disabled
- Single Instrument
- Shared Instrument

Or something like that?

File: Assets/Script/Settings/SettingsManager.Settings.cs [hunk 3]
```diff
@@ -155,8 +157,10 @@ public void OpenVenueFolder()
 
             public ToggleSetting ClapsInStarpower { get; } = new(true);
 
-            // public ToggleSetting UseWhammyFx            { get; } = new(true, UseWhammyFxChange);
-            // public SliderSetting WhammyPitchShiftAmount { get; } = new(1, 1, 12, WhammyPitchShiftAmountChange);
+            public ToggleSetting UseWhammyFx { get; } = new(false, UseWhammyFxChange);
```
[the remark this answers is not in the document]
Comment (jbienz) [reply Assets/Script/Settings/SettingsManager.Settings.cs:160]: I thought about this, but what would be the case for On (not Multitrack)? Wouldn't you be pitch bending the entire song, lyrics and all? Is that desired?

Or were you suggesting the same thing I was just with different naming? That we give an option to enable when more than one user shares the instrument?

File: Assets/StreamingAssets/lang/en-US.json [hunk 4]
```diff
@@ -596,6 +596,14 @@
                     "Off": "Off",
                     "On": "On"
                 }
+            },
+			"UseWhammyFx": {
+                "Name": "Use Whammy Fx",
+                "Description": "Enables pitch bending in response to the whammy bar on multi-track charts.\nWARNING: This setting is known to cause audio desync issues on some machines."
```
[the remark this answers is not in the document]
Comment (jbienz) [reply Assets/StreamingAssets/lang/en-US.json:602]: Maybe I'm missing the definition of multi-track? I thought multi-track meant stems, where guitar could be pitch bent separate from the rest of the track. If that's correct, I'm back to being confused about when we would want to support single track because I would think that would cause the entire song to pitch bend. Sorry if I'm missing something.

File: Assets/Script/Audio/Bass/BassStemChannel.cs [hunk 5]
```diff
@@ -35,23 +35,45 @@ internal BassStemChannel(AudioManager manager, SongStem stem, bool clampStemVolu
 
         protected override void SetWhammyPitch_Internal(float percent)
         {
-            if (_streamHandles.PitchFX == 0 || _reverbHandles.PitchFX == 0)
-                return;
-
             percent = Mathf.Clamp(percent, 0f, 1f);
```
[the remark this answers is not in the document]
Comment (jbienz) [reply Assets/Script/Audio/Bass/BassStemChannel.cs:38]: Okay. So I think you're saying don't clamp here just pass it on right?

File: Assets/Script/Settings/SettingsManager.Settings.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply Assets/Script/Settings/SettingsManager.Settings.cs:160]: `On` behaves like `Multitrack` if the player has its own stem. It only differs in that if they gets set to the default/background stem, you'd still be able to whammy it.

File: Assets/Script/Audio/Bass/BassStemChannel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply Assets/Script/Audio/Bass/BassStemChannel.cs:38]: correct

File: Assets/Script/Gameplay/GameManager.Audio.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply Assets/Script/Gameplay/GameManager.Audio.cs:195]: But again, you basically set up situations where players have their whammies being completely ignored purely based on someone else having priority (against their choice). The only way to have any sort of fairness would be to add some weighting or averaging, but these functions aren't built for that. You'd need to add functionality that sets the collective amounts of whammy percentage to zero on every frame. This also relies on the inputAction being sent every frame as well. If it is, then cool, but if not, then even that fails as stationary values won't be read on subsequent frames.

File: Assets/Script/Settings/SettingsManager.Settings.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply Assets/Script/Settings/SettingsManager.Settings.cs:160]: I guess `All` would be a better alternative.

File: Assets/StreamingAssets/lang/en-US.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sonicfind) [reply Assets/StreamingAssets/lang/en-US.json:602]: As mentioned, the third option would behave like multitrack *except* that the background stem could have whammy effects applied as well if an instrument ends up mapped to it.

**Comment by jbienz** [comment]
Replacing this PR with #846
